packages/periphery/contracts/oracles/UniswapV2Oracle.sol

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity 0.8.19;
+
+import '@openzeppelin/contracts/access/Ownable2Step.sol';
+import '@openzeppelin/contracts/utils/math/Math.sol';
+import '@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol';
+import '@uniswap/v2-periphery/contracts/libraries/UniswapV2OracleLibrary.sol';
+
+import '@marginly/contracts/contracts/interfaces/IPriceOracle.sol';
+
+/// @notice ExampleSlidingWindowOracle with a little modifications https://github.com/Uniswap/v2-periphery/blob/master/contracts/examples/ExampleSlidingWindowOracle.sol
+contract UniswapV2Oracle is IPriceOracle, Ownable2Step {
+  uint256 private constant X96ONE = 79228162514264337593543950336;
+
+  error WrongValue();
+  error WrongGranularity();
+  error WrongWindow();
+  error PairNotFound();
+  error MissingHistoricalObservation();
+  error UnexpectedTimeElapsed();
+  error PairAlreadyExists();
+
+  struct TokenPair {
+    address token0;
+    address token1;
+  }
+
+  struct Observation {
+    uint timestamp;
+    uint price0Cumulative;
+  }
+
+  struct PairOracleOptions {
+    uint16 secondsAgo;
+    uint16 secondsAgoLiquidation;
+  }
+
+  address public immutable factory;
+  // the desired amount of time over which the moving average should be computed, e.g. 24 hours
+  uint public immutable windowSize;
+  // the number of observations stored for each pair, i.e. how many price observations are stored for the window.
+  // as granularity increases from 1, more frequent updates are needed, but moving averages become more precise.
+  // averages are computed over intervals with sizes in the range:
+  //   [windowSize - (windowSize / granularity) * 2, windowSize]
+  // e.g. if the window size is 24 hours, and the granularity is 24, the oracle will return the average price for
+  //   the period:
+  //   [now - [22 hours, 24 hours], now]
+  uint8 public immutable granularity;
+  // this is redundant with granularity and windowSize, but stored for gas savings & informational purposes.
+  uint public immutable periodSize;
+
+  // mapping from pair address to a list of price observations of that pair
+  mapping(address => Observation[]) public pairObservations;
+  address[] public pairs;
```

Comment:
I suggest the next way of storing `token0IsBaseToken`: `bool token0IsBaseToken = token0IsBaseToken > 0; address pairAddress = address(uint160(abs(pairs[i])));` and extending `pairs[i]` type up to `int256` or something